more work based on cybereason: add logger and catch exceptions on network errors

realmgic opened the pull request.

realmgic committed:
fixing extend of the asinfo and asadm command (or else it fails). fixing the asadm command to run on all nodes using asinfo

realmgic committed:
remove old commands, commented out

realmgic committed:
handle when connection errors occur and assert the loop, add logger

realmgic committed:
Merge branch 'master' into master

realmgic committed:
Encoding the message to UTF-8 after script failed because of unicode to ascii error (on write, not logger but we handle it anyway here)

realmgic committed:
logging the beginning of the loop as info and not debug

realmgic committed:
Added more logging

GElkayam reviewed (approved):
Looking good.

realmgic commented:
@arrowplum can you please merge this? Thanks!

arrowplum commented:
I will leave some comments (thanks @kportertx ) but also going to add @spkesan as a reviewer

arrowplum commented on maintenance.py:
```diff
@@ -25,11 +25,31 @@
+METADATA_HEADERS = {'Metadata-Flavor': 'Google'}
+ASINFO = "/usr/bin/asinfo"
+ASADM = "/usr/bin/asadm"
+#AGM_LOG = "/var/log/aerospike/agm.log"
```
should remove

arrowplum commented on maintenance.py:
```diff
@@ -53,10 +68,18 @@ def wait_for_maintenance(callback):
-            url,
-            params={'last_etag': last_etag, 'wait_for_change': True},
-            headers=METADATA_HEADERS)
+        logger.debug("start loop")
```
this doesn't seem very helpful. a repeating "start loop" for each iteration.

realmgic replied:
the loop runs for an hour at a time (so this message printed every hour if still running), so this helped us understand when this is still running and when this failed.

arrowplum commented on maintenance.py:
```diff
@@ -81,25 +104,26 @@ def wait_for_maintenance(callback):
 def maintenance_callback(event):
     if event:
-        # print('Undergoing host maintenance: {}'.format(event))
+        logger.warning('Undergoing host maintenance: %s', event)
```
Should be info rather than warning, yeah?

realmgic replied:
this is as the customer asked - the thought process: if we're undergoing a maintenance event, this is "not normal" for the system (but is fine for the script) hence warning and not info.

spkesan replied:
From the script's perspective, it should still be INFO right. The script is meant to mitigate the impact during "live migration" or "termination" and starts to execute the logic as soon as (60 secs actually before the maintenance event starts) the `maintenance-event` attribute is set (like a response to an event). So the maintenance is scheduled already.
I may be wrong. Let me know your thoughts.

realmgic replied:
@spkesan you're right - but think of it from the customer perspective - it's either looking for a plain info saying "something bad is going on" (an event, meaning changes to his system) or to have it as bold "warning" stating the same.

the warning in that case is not that the script is doing something bad, is that the system is undergoing something bad...

spkesan replied:
Okay. A WARNING is good in that case.

arrowplum commented on maintenance.py:
```diff
@@ -81,25 +104,26 @@ def wait_for_maintenance(callback):
         asinfo1 = [ASINFO, "-v", "quiesce:"]
         asinfo1.extend(args.options.split())
         call(asinfo1)
+        logger.debug('quiesce finished')
```
shouldn't this also be info not debug?

realmgic replied:
yeah, i can change it

arrowplum commented on maintenance.py:
```diff
@@ -81,25 +104,26 @@ def wait_for_maintenance(callback):
 
     else:
-        # print('Finished host maintenance')
+        logger.warning('Finished host maintenance')
```
info not warning

realmgic replied:
see above about info vs. warning from system maintenance perspective

spkesan replied:
"Finished host maintenance" should be INFO.

arrowplum commented on maintenance.py:
```diff
@@ -81,25 +104,26 @@ def wait_for_maintenance(callback):
         asinfo2 = [ASINFO, "-v", "quiesce-undo:"]
         asinfo2.extend(args.options.split())
         call(asinfo2)
+        logger.debug('quiesce-undo finished')
```
info not debug

arrowplum commented on maintenance.py:
```diff
@@ -81,25 +104,26 @@ def wait_for_maintenance(callback):
     asadm.extend(args.options.split())
     call(asadm)
-
+    logger.debug('recluster finished')
```
info not debug

arrowplum reviewed (commented):
Just looked at the log lines. It seems like the level is all over the place. If I am wrong please let me know.

realmgic committed:
fixing some of the logger levels

spkesan commented on maintenance.py:
```diff
@@ -25,11 +25,31 @@
+METADATA_HEADERS = {'Metadata-Flavor': 'Google'}
+ASINFO = "/usr/bin/asinfo"
+ASADM = "/usr/bin/asadm"
+#AGM_LOG = "/var/log/aerospike/agm.log"
```
If using file logging, it would be better if we keep it as an absolute path (and standard one) `/var/log/aerospike/agm.log`. Need to uncomment it.

realmgic replied:
fixed

spkesan commented on maintenance.py:
```diff
@@ -25,11 +25,31 @@
+ASINFO = "/usr/bin/asinfo"
+ASADM = "/usr/bin/asadm"
+#AGM_LOG = "/var/log/aerospike/agm.log"
+AGM_LOG = "agm.log"
```
This has to be remove then, it would be better if we keep it as standard and absolute path.

realmgic replied:
fixed

spkesan commented on maintenance.py:
```diff
@@ -81,30 +110,33 @@ def wait_for_maintenance(callback):
         # realistically, any sort of maintenence event should drain aerospike
         asinfo1 = [ASINFO, "-v", "quiesce:"]
         asinfo1.extend(args.options.split())
         call(asinfo1)
```
It would be good to have these `asinfo` command output also sent to the `logger` so it's captured and it is also an important output for debugging in case of issues/RCA. 

You can do something like this to execute shell commands (Please correct any wrong syntax, I just wrote here directly):

```
def run_shell_command(command):
    p = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    stdout, stderr = p.communicate()
    if stdout:
        logger.info("Output: \n" + stdout)
    if stderr:
        logger.error("Error: \n" + stderr)

    if p.returncode != 0:
        logger.error("asinfo command: \"" + " ".join(asadm) + "\" returned error")
```
```
def maintenance_callback(event):
    if event:
        logger.warning('Undergoing host maintenance: %s', event)
        # realistically, any sort of maintenence event should drain aerospike
        asinfo1 = [ASINFO, "-v", "quiesce:"]
        asinfo1.extend(args.options.split())

        ## ---- Run quiesce ------
        logger.info('Running quiesce ..')
        run_shell_command(asinfo1)

    else:
        logger.warning('Finished host maintenance')
        asinfo2 = [ASINFO, "-v", "quiesce-undo:"]
        asinfo2.extend(args.options.split())

        ## ------ Run quiesce-undo ------
        logger.info('Running quiesce-undo ..')
        run_shell_command(asinfo2)

    asadm = [ASADM, "-e", "asinfo -v \"recluster:\""]
    asadm.extend(args.options.split())

    ## ------- Run recluster using asadm -------
    logger.info('Running recluster command across the cluster')
    run_shell_command(asadm)
    logger.info('recluster finished')
```

realmgic replied:
i wasn't sure we need to print the output if it's not failing - it is kind of spamming the log file and i wanted to try and avoid it. added the output good as debug and bad as error. 
wdyt?

spkesan replied:
Yeah. stdout as DEBUG makes sense.

spkesan reviewed (commented):
Thanks @realmgic for the PR.
I have added few comments and requested few changes. Let me know your thoughts.

realmgic committed:
fix log name

realmgic committed:
added printing of output if error to the log.
if output is good, i don't think we need to overload the log (but we can print on debug)

spkesan commented on maintenance.py:
```diff
@@ -38,13 +57,19 @@
-METADATA_HEADERS = {'Metadata-Flavor': 'Google'}
-ASINFO = "/usr/bin/asinfo"
-ASADM = "/usr/bin/asadm"
+def run_shell_command(command):
```
It would be better to handle the `stderr` first and `return`. Also need a INFO indication that the command ran successfully.
```
def run_shell_command(command):
    p = subprocess.Popen(command, stdout=subprocess.PIPE,
                         stderr=subprocess.PIPE)
    stdout, stderr = p.communicate()
    if stderr:
        logger.error("command: \"" + " ".join(command) +
                     "\" Error: \n" + stderr)
        return

    if stdout:
        logger.debug("command: \"" + " ".join(command) +
                     "\" Output: \n" + stdout)
    if p.returncode != 0:
        logger.error("Command: \"" + " ".join(command) +
                     "\" returned with error code: " + str(p.returncode))
    else:
        logger.info("Command \"" + " ".join(command) + "\" ran successfully")
```

If printing the "command" here, then we can remove the `logger.debug('options: %s', args.options)` line above.

realmgic replied:
the options %s are the args for the entire script, not just this run shell command function (i.e. user-password, TLS, etc.)

spkesan replied:
There's only one option in the script to specify the options for the asinfo or asadm commands.
For future use, makes sense.

spkesan commented on maintenance.py:
```diff
@@ -53,18 +78,31 @@ def wait_for_maintenance(callback):
+                headers=METADATA_HEADERS)
+            logger.info("agm status change, running... new status code: %s. text: %s", r.status_code, r.text.encode('utf-8').strip())
+
+        except requests.exceptions.ConnectionError as err:
```
Remove `err` if not using.

We should retry for all exceptions from `requests.get()`
https://2.python-requests.org//en/latest/user/quickstart/#errors-and-exceptions
Added comments for why 503, and all status codes from metadata server.
https://cloud.google.com/compute/docs/storing-retrieving-metadata#statuscodes
```
        try:
            r = requests.get(
                url,
                params={'last_etag': last_etag, 'wait_for_change': False},
                headers=METADATA_HEADERS)
            logger.info("agm status change, running... new status code: %s. text: %s",
                        r.status_code, r.text.encode('utf-8').strip())

        except requests.exceptions.RequestException:
            # Retry for any exception from requests.get()
            # Check for 503 and raise_for_status() - 4XX and 5XX will be done separately.
            # https://2.python-requests.org//en/latest/user/quickstart/#errors-and-exceptions
            time.sleep(1)
            continue

        # During maintenance the service can return a 503, so these should
        # be retried.
        # https://cloud.google.com/compute/docs/storing-retrieving-metadata#statuscodes
        if r.status_code == 503:
            time.sleep(1)
            continue

        # Any other response from metadata service will kill this script
        try:
            r.raise_for_status()
        except requests.exceptions.HTTPError as error:
            logger.error(
                "metadata service returned bad code, terminating! (%s)", str(error))
            raise error
```

realmgic replied:
fixed

spkesan commented on maintenance.py:
```diff
@@ -53,18 +78,31 @@ def wait_for_maintenance(callback):
+            r.raise_for_status()
+	except requests.exceptions.HTTPError as error:
+	    logger.error("metadata service returned bad code, terminating! (%s)", str(error))
+	    raise
```
`raise error`

realmgic replied:
fixed

spkesan commented on maintenance.py:
```diff
@@ -81,30 +119,33 @@ def wait_for_maintenance(callback):
+        logger.warning('Undergoing host maintenance: %s', event)
         # realistically, any sort of maintenence event should drain aerospike
         asinfo1 = [ASINFO, "-v", "quiesce:"]
         asinfo1.extend(args.options.split())
```
Can you also rename `asinfo1` and `asinfo2` to be just `asinfo`.

```
    if event:
        logger.warning('Undergoing host maintenance: %s', event)
        # realistically, any sort of maintenence event should drain aerospike
        asinfo = [ASINFO, "-v", "quiesce:"]
        asinfo.extend(args.options.split())
        run_shell_command(asinfo)
        logger.info('quiesce finished')

    else:
        logger.info('Finished host maintenance')
        asinfo = [ASINFO, "-v", "quiesce-undo:"]
        asinfo.extend(args.options.split())
        run_shell_command(asinfo)
        logger.info('quiesce-undo finished')
```

realmgic replied:
i know, it's just to make sure we don't step on the variable (we had such a bug in the code before - with the asinfo and the asadm the comes later). is it important to reuse the name?

realmgic committed:
fix run_shell_command prints and order

realmgic committed:
fixing the error handling for the metadata request

spkesan committed:
Fix exception handling added comments, plus minor fixes
